### PR TITLE
connection: run all pseudoconsole hosts in jobs

### LINKED ISSUE #1250: Windows Terminal hangs when shell exits
# Environment

```none
Windows build number: Microsoft Windows [Version 10.0.18362.175]
Windows Terminal version (if applicable): d82eab44d076e441f0b630d9b53a19dfa884c3ab
```

# Steps to reproduce

1. Open Windows Terminal
2. Have your shell terminate, e.g. by typing `exit`

# Expected behavior

The tab, or the whole window, closes.

# Actual behavior

Windows Terminal hangs, leading to a _WindowsTerminal.exe is not responding_ eventually.

### LINKED ISSUE #1275: [Bug Report] crash after type command 'exit'
# Environment

```none
Windows build number:  10.0.18917.1000
Windows Terminal version: commit 315abf6fa6e096f0ad8ba7db68fa8a3fd9a591b1
```

# Steps to reproduce

1. start Terminal application
2. run some commands in the prompt
3. run command `exit`

# Expected behavior

Simply exit shell.

# Actual behavior

image

Run `exit` command straightly didn't cause a crash, but it will happen after run any other command. It happens in all the default `PowerShell`, `cmd`, `WSL` profiles.

the profiles JSON file's modifications are:

```diff
--- profiles.old.json	2019-06-15 17:07:10.565458400 +0800
+++ profiles.json	2019-06-15 17:07:06.089459800 +0800
@@ -1,10 +1,10 @@
 {
     "globals" :
     {
-        "alwaysShowTabs" : true,
+        "alwaysShowTabs" : false,
         "defaultProfile" : "{61c54bbd-c2c6-5271-96e7-009a87ff44bf}",
-        "initialCols" : 120,
-        "initialRows" : 30,
+        "initialCols" : 80,
+        "initialRows" : 24,
         "keybindings" :
         [
             {
@@ -199,28 +199,28 @@
         ],
         "requestedTheme" : "system",
         "showTabsInTitlebar" : true,
-        "showTerminalTitleInTitlebar" : true
+        "showTerminalTitleInTitlebar" : false
     },
     "profiles" :
     [
         {
-            "acrylicOpacity" : 0.5,
-            "background" : "#012456",
+            "acrylicOpacity" : 0.30000001192092896,
+            "background" : "#282C34",
             "closeOnExit" : true,
-            "colorScheme" : "Campbell",
+            "colorScheme" : "One Half Dark",
             "commandline" : "powershell.exe",
             "cursorColor" : "#FFFFFF",
             "cursorShape" : "bar",
-            "fontFace" : "Consolas",
+            "fontFace" : "Sarasa Term SC",
             "fontSize" : 10,
             "guid" : "{61c54bbd-c2c6-5271-96e7-009a87ff44bf}",
             "historySize" : 9001,
             "icon" : "ms-appx:///ProfileIcons/{61c54bbd-c2c6-5271-96e7-009a87ff44bf}.png",
             "name" : "Windows PowerShell",
-            "padding" : "0, 0, 0, 0",
+            "padding" : "4, 4, 4, 4",
             "snapOnInput" : true,
             "startingDirectory" : "%USERPROFILE%",
-            "useAcrylic" : false
+            "useAcrylic" : true
         },
         {
             "acrylicOpacity" : 0.75,
```

### PR DESCRIPTION
**Pull request by DHowett-MSFT**
This commit also switches the manual resource management in
ConhostConnection to use WIL.

This change also starts conhost suspended so that we can get our lives in order before we let it start.

This seems like good practice, and it will almost certainly help with the orphaning of some conhosts after WT exits.

## PR Checklist
* [x] Closes #1275, Closes #1250
* [x] CLA signed
* [x] Tests added/passed/not required
* [x] Documentation updated or not required
* [x] I've discussed this with core contributors already: this was a backlog item from before GitHub

### CODE REVIEW

File: src/cascadia/TerminalConnection/ConhostConnection.cpp [hunk 1]
```diff
@@ -29,7 +29,8 @@ namespace winrt::Microsoft::Terminal::TerminalConnection::implementation
         _initialCols{ initialCols },
         _commandline{ commandline },
         _startingDirectory{ startingDirectory },
-        _guid{ initialGuid }
+        _guid{ initialGuid },
+        _connected{ false }
```
Comment (DHowett-MSFT) [inline src/cascadia/TerminalConnection/ConhostConnection.cpp]: @binarycrusader I may have gotten the outcome of the "where do I set values on construct?" discussion backwards. :smile:

**Comment by DHowett-MSFT** [comment]
Okay, this is in flagrant violation of _C.48 Prefer in-class initializers to member initializers in constructors for constant initializers._ I got the rules backwards. :smile:

File: src/cascadia/TerminalConnection/ConhostConnection.h [hunk 2]
```diff
@@ -26,20 +26,22 @@ namespace winrt::Microsoft::Terminal::TerminalConnection::implementation
...
-        guid _guid{}; // A "unique" session identifier for connected client
-        bool _closing{};
+        // set during construction - not constant
+        uint32_t _initialRows;
+        uint32_t _initialCols;
+        hstring _commandline;
+        hstring _startingDirectory;
+        guid _guid; // A "unique" session identifier for connected client
+
+        bool _connected{ false };
```
Comment (DHowett-MSFT) [inline src/cascadia/TerminalConnection/ConhostConnection.h]: I realize that `{}` is equivalent to `{false}`, but it seems important to me to say `{false}` or `{true}` because a boolean has only two states.

File: src/cascadia/TerminalConnection/ConhostConnection.cpp [hunk 3]
```diff
@@ -92,31 +92,51 @@ namespace winrt::Microsoft::Terminal::TerminalConnection::implementation
...
                          extraEnvVars));
 
-        _connected = true;
+        _hJob.reset(CreateJobObjectW(nullptr, nullptr));
+        THROW_LAST_ERROR_IF_NULL(_hJob);
+
+        // We want the conhost and all associated descendant processes
+        // to be terminated when the tab is closed. GUI applications
+        // spawned from the shell tend to end up in their own jobs.
+        JOBOBJECT_EXTENDED_LIMIT_INFORMATION jobExtendedInformation{ };
```
Comment (binarycrusader) [inline src/cascadia/TerminalConnection/ConhostConnection.cpp]: nit: spaces between {} init braces or not?  Don't care, would just like to know what the preferred style is.

File: src/cascadia/TerminalConnection/ConhostConnection.h (hunk 2 above)
Comment (binarycrusader) [reply src/cascadia/TerminalConnection/ConhostConnection.h]: `{}` implies default value initialization, so yes, it is the same.  Personally, the only reason I prefer {} over {false}, is because when there a large number of class members declared together, {true} sticks out very obviously:

```C++
bool member1{};
bool member1{};
bool member2{};
bool member3{true};
bool member4{};
```

However, it's a matter of personal preference, so whatever the project prefers is fine.

File: src/cascadia/TerminalConnection/ConhostConnection.h [hunk 4]
```diff
@@ -26,20 +26,22 @@ namespace winrt::Microsoft::Terminal::TerminalConnection::implementation
...
-        bool _connected{};
-        HANDLE _inPipe{ INVALID_HANDLE_VALUE };  // The pipe for writing input to
-        HANDLE _outPipe{ INVALID_HANDLE_VALUE }; // The pipe for reading output from
-        HANDLE _signalPipe{ INVALID_HANDLE_VALUE };
-        DWORD _outputThreadId{};
-        HANDLE _hOutputThread{ INVALID_HANDLE_VALUE };
-        PROCESS_INFORMATION _piConhost{};
-        guid _guid{}; // A "unique" session identifier for connected client
-        bool _closing{};
+        // set during construction - not constant
```
Comment (binarycrusader) [inline src/cascadia/TerminalConnection/ConhostConnection.h]: Even though these are initialized in the constructor, there's no harm in also initializing them here.  The compiler is smart enough to do the right thing.

Personally, I prefer to initialize them in both places because it ensures that members of the class are never in an invalid or unknown state.  For example, if the members are initialized in the class constructor, it ensures they have a sane default value until they are explicitly initialized in the constructor.  In addition, it also means that if someone removes the initialization from the constructor later, there will still be a good default value.

With that said, if this is what the project prefers, that's fine.

File: src/cascadia/TerminalConnection/ConhostConnection.cpp (hunk 3 above)
Comment (adiviness) [reply src/cascadia/TerminalConnection/ConhostConnection.cpp]: The style is `{}` if there's no params and spaces if there are, like `{ foo, bar }`

File: src/cascadia/TerminalConnection/ConhostConnection.h (hunk 4 above)
Comment (dlong11) [reply src/cascadia/TerminalConnection/ConhostConnection.h]: I agree with @binarycrusader.   This "rule"  - makes it easier for the developer to get the original guideline right and it makes the code review much simpler.

Comment (DHowett-MSFT) [reply src/cascadia/TerminalConnection/ConhostConnection.h]: So here's a question: for things that have a meaningful default constructor, do we still put `{}` in the class? For POD types it's obviously required, but for `winrt::hstring` and `wil::unique_x` it's implicit and it's always initialized properly. Thoughts?

Comment (dlong11) [reply src/cascadia/TerminalConnection/ConhostConnection.h]: @DHowett-MSFT In my day job, we don't add {} for classes/objects.  
Our rule has more detail with proper "c++ legalese"  but it is basicallly
Use {} for all POD,  structs.  (i.e. - any object that doesn't get initialized at declaration)
For enums and enum class we require that you provide a default enum value.  (enums with {} have the potential to be completely wrong)

Since structs can be default initialized correctly with member initializing you could argue that they don't need to use {}.   But again to keep this simple we require the use of {} because most structs do not member initialize their variables.

Not sure if this is right, but seeing objects that have ctors that have {} seems wrong to us.

Comment (dlong11) [reply src/cascadia/TerminalConnection/ConhostConnection.h]: Note - you can have execptions like winrt::hstring.  It is a struct but you can treat it like an object because we are 100% sure it will be initialized correctly.  (I don't have the source on hand for winrt::string so I am asumming this based on your comments)

Comment (dlong11) [reply src/cascadia/TerminalConnection/ConhostConnection.h]: Looking at winrt::hstring closer - if you were to have a class member that is an hstring then it should be declared with {}.  Since it is a plain struct.

Comment (binarycrusader) [reply src/cascadia/TerminalConnection/ConhostConnection.h]: > So here's a question: for ~things~ members that have a meaningful default constructor, do we still put `{}` in the class? For POD types it's obviously required, but for `winrt::hstring` and `wil::unique_x` it's implicit and it's always initialized properly. Thoughts?

My historical approach has been that non-POD objects (classes) never get {} applied to them since the default constructor will be executed when the parent object is constructed because I trust the constructor of the class to initialize it.  I then applied {} to all PODs and structs because equally because I didn't want to have to consider whether they a constructor defined.

I think the guidance the ISO C++ Guidelines provide is simplest here:

[C.48: Prefer in-class initializers to member initializers in constructors for constant initializers](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c48-prefer-in-class-initializers-to-member-initializers-in-constructors-for-constant-initializers)

In short, if you look at the example there, the guidance seems to suggest that all members that should have an initial, constant value should have an in-class initializer and that the constructors for the class should then use member initializers in the constructors in member definition order.

Unfortunately, I don't think there's any "right, obvious answer" here.  So my personal suggestion would be to look at the relevant guideline, pick whatever interpretation is simplest to apply and go with that.  As long as it's consistent, I think everyone will be happy.

Comment (binarycrusader) [reply src/cascadia/TerminalConnection/ConhostConnection.h]: > For enums and enum class we require that you provide a default enum value. (enums with {} have the potential to be completely wrong)

Personally, I have a similar, but slightly different approach.

enum or enum class members are always declared with the initial enum value (0) being an 'Unknown' or 'Invalid' value.  For classes and structs that have those enum members where I expect a value to be provided upon construction, the in-class member initializers have {}, which makes it easy to spot objects that were not constructed correctly since no valid struct/class should ever have those values for the enums.

Comment (dlong11) [reply src/cascadia/TerminalConnection/ConhostConnection.h]: Interesting.  I have to run but I might ping you later and ask a few questions around this.  Always looking for better guidelines.   Thanks!

File: src/cascadia/TerminalConnection/ConhostConnection.h (hunk 2 above)
Comment (miniksa) [reply src/cascadia/TerminalConnection/ConhostConnection.h]: I personally prefer `{false}` because it subtracts the mental load of evaluating the default for a type in my head.

Comment (binarycrusader) [reply src/cascadia/TerminalConnection/ConhostConnection.h]: I sort of understand, but the default for every type though is effectively zero.  That doesn’t seem very mysterious to me?

File: src/cascadia/TerminalConnection/ConhostConnection.cpp (hunk 3 above)
Comment (DHowett-MSFT) [reply src/cascadia/TerminalConnection/ConhostConnection.cpp]: The style has come through and fixed it!

File: src/cascadia/TerminalConnection/ConhostConnection.h (hunk 4 above)
Comment (DHowett-MSFT) [reply src/cascadia/TerminalConnection/ConhostConnection.h]: I ended up choosing to use `{}` for POD and POD-likes, and standard default construction for non-POD things.

File: src/cascadia/TerminalConnection/ConhostConnection.h (hunk 2 above)
Comment (DHowett-MSFT) [reply src/cascadia/TerminalConnection/ConhostConnection.h]: I chose `{}` except for the atomic, which requires initialization.
